### PR TITLE
Add Prometheus Instrumentation

### PR DESCRIPTION
**Pull request by ghost**
This PR introduces some scaffolding for [Prometheus](https://prometheus.io/) instrumentation to the project and adds a some key metrics to the indexing pipeline.

__Test__

Run the instrumentation (Prometheus + Grafana) infrastructure locally by following the README in the [instrumentation](https://github.com/graphprotocol/instrumentation) repo.

Deploy a subgraph to a local `graph-node` running on the source from this branch and visit `http://localhost:3000` (Grafana) to look at the `Subgraph Metrics` dashboard.

### CODE REVIEW

File: server/metrics/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        PrometheusMetricsServeError::BindError(err)
+    }
+}
+
+pub struct PrometheusMetricsServer {
+    logger: Logger,
+    registry: Arc<Registry>,
+}
+
+impl std::clone::Clone for PrometheusMetricsServer {
```
Comment (Jannis) [inline server/metrics/src/lib.rs]: I don't think you need to import `std::clone::Clone`, just `Clone` could be enough. Either way, I'd just have `Clone` here and if we need to import anything, I'd do that at the top. Same with `fmt::Display` above.

**Review by Jannis — CHANGES_REQUESTED** [review]

File: core/src/subgraph/instance_manager.rs [hunk 2]
```diff
@@ -48,25 +50,154 @@ where
...
+
 impl SubgraphInstanceManager {
     /// Creates a new runtime manager.
-    pub fn new<B, S, T>(
+    pub fn new<B, S, T, M>(
         logger_factory: &LoggerFactory,
         stores: HashMap<String, Arc<S>>,
         host_builder: T,
         block_stream_builder: B,
+        registry: Arc<M>,
```
[the remark this answers is not in the document]
Comment (olibearo) [reply core/src/subgraph/instance_manager.rs]: Renamed, but was unable to use `<dyn MetricsRegistry>`. I got this error:

```
the trait `graph::prelude::MetricsRegistry` cannot be made into an object
```

File: core/src/subgraph/instance_manager.rs [hunk 3]
```diff
@@ -242,9 +400,14 @@ impl SubgraphInstanceManager {
...
         //
         // This task has many calls to the store, so mark it as `blocking`.
-        tokio::spawn(graph::util::futures::blocking(loop_fn(ctx, |ctx| {
-            run_subgraph(ctx)
-        })));
+        let subgraph_runner =
+            graph::util::futures::blocking(loop_fn(ctx, move |ctx| run_subgraph(ctx))).then(
+                move |res| {
+                    subgraph_metrics_unregister.unregister(registry);
+                    future::result(res)
```
[the remark this answers is not in the document]
Comment (olibearo) [reply core/src/subgraph/instance_manager.rs]: That will work in the case where `run_subgraph` returns `Ok`, but not in the `Err` case

**Comment by Jannis** [comment]
🍕

File: core/src/subgraph/instance_manager.rs [hunk 4]
```diff
@@ -48,25 +50,154 @@ where
...
+    pub fn inc_subgraph_count(&self) {
+        self.subgraph_gauge.inc();
+    }
+
+    pub fn dec_subgraph_count(&self) {
+        self.subgraph_gauge.dec();
+    }
+}
+
+enum TriggerType {
```
[the remark this answers is not in the document]
Comment (olibearo) [reply core/src/subgraph/instance_manager.rs]: Yeah, but the types include the trigger data in the enum tuple and I did not want to pass those around

File: core/src/subgraph/instance_manager.rs [hunk 5]
```diff
@@ -156,25 +295,28 @@ impl SubgraphInstanceManager {
...
 
-    fn start_subgraph<B, T, S>(
+    fn start_subgraph<B, T, S, M>(
         logger: Logger,
         instances: SharedInstanceKeepAliveMap,
         host_builder: T,
         stream_builder: B,
         store: Arc<S>,
         manifest: SubgraphManifest,
+        registry: Arc<M>,
```
[the remark this answers is not in the document]
Comment (olibearo) [reply core/src/subgraph/instance_manager.rs]: This didn't end up working

File: core/src/subgraph/instance_manager.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Jannis) [reply core/src/subgraph/instance_manager.rs]: Ok, that's fair enough.

File: core/src/subgraph/instance_manager.rs [hunk 6]
```diff
@@ -80,26 +212,29 @@ impl SubgraphInstanceManager {
...
         stores: HashMap<String, Arc<S>>,
         host_builder: impl RuntimeHostBuilder,
         block_stream_builder: B,
+        metrics_registry: Arc<M>,
     ) where
         S: Store + ChainStore + SubgraphDeploymentStore + EthereumCallCache,
         B: BlockStreamBuilder,
+        M: MetricsRegistry,
     {
         // Subgraph instance shutdown senders
```
Comment (Jannis) [inline core/src/subgraph/instance_manager.rs]: This comment is only relevant to the `let instances` below, right? Shall we separate that from the metrics variables?

File: core/src/subgraph/instance_manager.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Jannis) [reply core/src/subgraph/instance_manager.rs]: Ok, no problem

File: graph/src/metrics/registry.rs [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+            namespace,
+        }
+    }
+
+    pub fn register(&self, name: String, c: Box<dyn Collector>) {
+        let err = match self.registry.register(c).err() {
+            None => return,
+            Some(err) => err,
+        };
+        match err {
```
[the remark this answers is not in the document]
Comment (olibearo) [reply graph/src/metrics/registry.rs]: Addressed this by adding metrics to the `MetricsRegistry`. I still want to protect the process from crashes due to metrics registration errors.

**Review by Jannis — APPROVED** [review]

**Pull request merged**